### PR TITLE
Solutions

### PR DESCRIPTION
**Pull request by Ramizjon**
Added solution of third task

### CODE REVIEW

File: .classpath [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (btrofimov) [inline .classpath]: @Ramizjon  - I would suggest to exclude these eclipse-based files from the project, the same about `.project`,`.settings/*`
The best option is to exclude them from git and add special rule to `.gitignore` file.

File: Problem1/src/StringChecker.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import java.util.HashSet;
+import java.util.Set;
+
+// TASK:
+// Implement an algorithm to determine if a string has all unique characters. 
+// What if you cannot use additional data structures?
+
+
+public class StringChecker {
```
Comment (btrofimov) [inline Problem1/src/StringChecker.java]: Please use `javadoc` comments in order to explain task and your made decisions.

File: Problem1/src/UnitTest.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import static org.junit.Assert.*;
+
+import org.junit.Test;
+
+
+public class UnitTest {
+
+	@Test
+	public void testAllUnique1 (){
```
Comment (btrofimov) [inline Problem1/src/UnitTest.java]: good, one important rule of designing unit tests is to cover edge cases like ampty strings etc. You have to make work every route inside your `checkAllUniqueChars1``

One important note. Instead of using things like checkAllUniqueChars1 ,checkAllUniqueChars2 - pelase add common itnerface and just provide two implementations via corresponding classes.

File: Problem2/src/StringComparator.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+public class StringComparator {
+	public boolean compareStrings (String firstStr, String secondStr){
+		char [] firstStringChars = firstStr.toCharArray();
+		char [] secondStringChars = secondStr.toCharArray();
+		Arrays.sort(firstStringChars);
+		Arrays.sort(secondStringChars);
+		
+		//simply compare sorted Strings for elements equality
+		if (Arrays.equals(firstStringChars, secondStringChars))
```
Comment (btrofimov) [inline Problem2/src/StringComparator.java]: Would it be better to use something like this

```
return Arrays.equals(firstStringChars, secondStringChars);
```

File: Problem2/src/StringComparator.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+			return true;
+		
+		return false;
+	}
+}
```
Comment (btrofimov) [inline Problem2/src/StringComparator.java]: Please provide unit tests

File: Problem3/src/StringCompressor.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+		char [] charsFromString = inputString.toCharArray();
+		StringBuilder sb = new StringBuilder ();
+
+		int counter = 1;
+		for (int i = 0; i < charsFromString.length; i++){
+			if ((i!=charsFromString.length-1) && (charsFromString[i]==charsFromString[i+1])){
+				counter++;
+			}
+			else {
+				sb.append(String.valueOf(charsFromString[i]));
```
Comment (btrofimov) [inline Problem3/src/StringCompressor.java]: I suspect there should be overriden method accepting Chars instead of `String.valueOf(`

File: Problem3/src/StringCompressor.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+				sb.append(String.valueOf(counter));
+				counter=1;
+			}
+		}
+		return (inputString.length()>sb.length())?sb.toString():inputString;
+	}
+}
```
Comment (btrofimov) [inline Problem3/src/StringCompressor.java]: Good implementation. Two general comments. Pelase provide unit tests.
Teh second comment. Let's re-implement this task with java streams as one more implementation (the same idea - via java interface and several impementations)

File: Problem4/src/SpacesReplacer.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+		for (int i = 0; i < inputString.length; i++){
+			if (inputString[i]==' '){
+				resultString[resultIndex++]='%';
+				resultString[resultIndex++]='2';
+				resultString[resultIndex++]='0';
+			}
+			else 
+				resultString[resultIndex++]=inputString[i];
+		}
+		return new String (resultString);
```
Comment (btrofimov) [inline Problem4/src/SpacesReplacer.java]: Lets reimplement this task with java streams. and where are your unit tests? ;)

**Comment by Ramizjon** [comment]
@btrofimov i have fixed some small issues that were present in previous pull request

**Comment by btrofimov** [comment]
@Ramizjon - I stil lsee build-specific files like
.classpath
.metadata/*
.settings/*
 .bash_profile,
Make sure that all nested directories do not contain these files as well.

**Comment by btrofimov** [comment]
Also I still do not see javadoc-like comments, Please use this tutorial http://www.oracle.com/technetwork/articles/java/index-137868.html and let's re-organize your comments inthat way.

**Comment by Ramizjon** [comment]
@btrofimov added javadoc comments and made .gitignore to work as you requested.

**Comment by btrofimov** [comment]
One more time, please make sure that you repository contains _only_ source code, NOT eclipse-based stuff.
I do see `.recommenders`  `Probblem5/target/classes` folder  inside PR

**Comment by Ramizjon** [comment]
@btrofimov  could you please checkout files now

File: Problem1/src/StringChecker2.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
+import java.util.HashSet;
+import java.util.Set;
+
+
+
+
+public class StringChecker1 implements StringChecker {
+
+	//Solution with use of Set as an additional data structure
```
Comment (btrofimov) [inline Problem1/src/StringChecker2.java:9]: Not a javadoc srtle comment

**Comment by btrofimov** [comment]
well then, much better for now. I think you can merge it  and delete corresponding branch from github directly.
